### PR TITLE
mgr/cephadm: Move `orch daemon add` to cephadm namespace

### PR DESCRIPTION
**Pull request by sebastian-philipp**
* Clearly state those commands ignore most checks

Fixes: https://tracker.ceph.com/issues/44673

**Blocked by**

- [x] #35087

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -1686,7 +1686,9 @@ def matches_network(host):
         r = False
 
         if daemon_type == 'osd':
-            return False if create_func(spec) else True # type: ignore
+            # TODO: return True would result in a busy loop
+            create_func(spec)
+            return False
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:1691]: @bk201 what do you think?

Comment (mgfritch) [reply src/pybind/mgr/cephadm/module.py:1642]: `return False` is confusing ... can we add a log line or comment as to why ...?

File: src/pybind/mgr/orchestrator/module.py [hunk 2]
```diff
@@ -706,7 +672,7 @@ def _mds_add(self,
             placement=PlacementSpec.from_string(placement),
         )
 
-        completion = self.add_mds(spec)
```
Comment (mgfritch) [inline src/pybind/mgr/orchestrator/module.py:709]: remove this func? duplicated by `cephadm daemon add` ?

File: src/pybind/mgr/orchestrator/module.py [hunk 3]
```diff
@@ -740,7 +706,7 @@ def _rgw_add(self,
             placement=PlacementSpec.from_string(placement),
         )
 
-        completion = self.add_rgw(spec)
```
Comment (mgfritch) [inline src/pybind/mgr/orchestrator/module.py:743]: remove this func? duplicated by `cephadm daemon add` ?

File: src/pybind/mgr/orchestrator/module.py [hunk 4]
```diff
@@ -769,7 +735,7 @@ def _nfs_add(self,
         )
 
         spec.validate_add()
-        completion = self.add_nfs(spec)
```
Comment (mgfritch) [inline src/pybind/mgr/orchestrator/module.py:772]: remove this func? duplicated by `cephadm daemon add` ?

File: src/pybind/mgr/orchestrator/module.py [hunk 5]
```diff
@@ -795,7 +761,7 @@ def _iscsi_add(self,
             placement=PlacementSpec.from_string(placement),
         )
 
-        completion = self.add_iscsi(spec)
```
Comment (mgfritch) [inline src/pybind/mgr/orchestrator/module.py:798]: remove this func? duplicated by `cephadm daemon add` ?

File: PendingReleaseNotes [hunk 6]
```diff
@@ -367,3 +367,8 @@
   the cluster, which could fill a nearly-full cluster.  They have been
   replaced by a tool, currently considered experimental,
   ``rgw-orphan-list``.
+
+* Cephadm: ``ceph orch daemon add`` was renamed to
+  ``ceph cephadm remote daemon add`` to clarify that those commands work
```
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply PendingReleaseNotes:372]: I was searching for something that is we're not simply adding a new daemon to a service, but instead directly deploy a daemon without looking for co-location conflicts etc.

File: src/pybind/mgr/orchestrator/_interface.py [hunk 7]
```diff
@@ -889,6 +889,34 @@ def next(ls):
...
+                raise OrchestratorValidationError(f"service placement not explicit {placement}")
+            if not new_spec.placement.is_explicit():
+                raise OrchestratorValidationError(f"new placement not explicit {new_spec.placement}")
+            placement = placement.copy()
+            placement.hosts = list(HostPlacementSpec.union(new_spec.placement.hosts, placement.hosts))
+            if new_spec.placement.count is not None:
+                if placement.count:
+                    placement.count += new_spec.placement.count
+                else:
+                    placement.count = new_spec.placement.count
```
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/_interface.py:904]: no idea what's the bets thing to do here. 

imagine `placement == 'count:2 host1 host2 host3' and `new_spec.placement == `count:1 host4`. What does that even mean.

**Comment by sebastian-philipp** [comment]
```
ceph/deployment/drive_group.py:304:1: W391 blank line at end of file
ceph/deployment/service_spec.py:11:1: E302 expected 2 blank lines, found 1
ceph/deployment/service_spec.py:137:5: E303 too many blank lines (2)
```

**Comment by sebastian-philipp** [comment]
@mgfritch thanks for the review! 

> generally don't we want to remove all the `orch daemon add` commands in favor of the new `cephadm daemon add` commands in this PR?

That's a usability thing. Right now, there is no single command to add a new daemon to a service without either 

1. exporting the service 
2. edit the spec
3. apply the spec

or painfully assemble a command line that does the same thing on the cli.

Having a way to simply extend a service was indeed requested more than once.

**Comment by jschmid1** [comment]
> generally don't we want to remove all the `orch daemon add` commands in favor of the new `cephadm daemon add` commands in this PR?

I'd like to see it disappear from the _orch_ cli as well.. 

> Having a way to simply extend a service was indeed requested more than once.

which comes with a lot more downsides that the people that are requesting this anticipate..

The more ways we add/support to add daemons the more weirdness we'll have to deal with.

**Comment by sebastian-philipp** [comment]
> > Having a way to simply extend a service was indeed requested more than once.
> 
> which comes with a lot more downsides that the people that are requesting this anticipate..
> 
> The more ways we add/support to add daemons the more weirdness we'll have to deal with.

Right. k8s also won't give you a way to simply add a new host. Instead they focus of setting the number of replica count.

**Comment by jschmid1** [comment]
> > > Having a way to simply extend a service was indeed requested more than once.
> > 
> > 
> > which comes with a lot more downsides that the people that are requesting this anticipate..
> > The more ways we add/support to add daemons the more weirdness we'll have to deal with.
> 
> Right. k8s also won't give you a way to simply add a new host. Instead they focus of setting the number of replica count.

And I'd absolutely follow that approach.

**Comment by sebastian-philipp** [comment]
> > > > Having a way to simply extend a service was indeed requested more than once.
> > > 
> > > 
> > > which comes with a lot more downsides that the people that are requesting this anticipate..
> > > The more ways we add/support to add daemons the more weirdness we'll have to deal with.
> > 
> > 
> > Right. k8s also won't give you a way to simply add a new host. Instead they focus of setting the number of replica count.
> 
> And I'd absolutely follow that approach.

but then we really need to improve the UX for this first, I think.

File: src/pybind/mgr/cephadm/module.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:1642]: moved to #35087

**Comment by sebastian-philipp** [comment]
But in general, I'm against removing `orch daemon add` without any notice period. cephadm is already out there and people use it. IMO we have to keep the stability there. For pacific, we can remove it then.

**Comment by mgfritch** [comment]
> > > > > Having a way to simply extend a service was indeed requested more than once.
> > > > 
> > > > 
> > > > which comes with a lot more downsides that the people that are requesting this anticipate..
> > > > The more ways we add/support to add daemons the more weirdness we'll have to deal with.
> > > 
> > > 
> > > Right. k8s also won't give you a way to simply add a new host. Instead they focus of setting the number of replica count.
> > 
> > 
> > And I'd absolutely follow that approach.
> 
> but then we really need to improve the UX for this first, I think. like e.g.e
> 
> https://kubernetes.io/docs/reference/kubectl/cheatsheet/#scaling-resources

It makes much more sense to separate concerns here -
* `apply` for create/update of the ServiceSpec
* `scale` for modifications to the PlacementSpec count

which implies `--placement` should be dropped from the `orch apply` cli ...

**Comment by jschmid1** [comment]
this brings us back to https://github.com/ceph/ceph/pull/35084#discussion_r425856138 

let's schedule this for todays orchestrator meeting

**Comment by sebastian-philipp** [comment]
rebased.

> > > > > > Having a way to simply extend a service was indeed requested more than once.
> > > > > 
> > > > > 
> > > > > which comes with a lot more downsides that the people that are requesting this anticipate..
> > > > > The more ways we add/support to add daemons the more weirdness we'll have to deal with.
> > > > 
> > > > 
> > > > Right. k8s also won't give you a way to simply add a new host. Instead they focus of setting the number of replica count.
> > > 
> > > 
> > > And I'd absolutely follow that approach.
> > 
> > 
> > but then we really need to improve the UX for this first, I think. like e.g.e
> > https://kubernetes.io/docs/reference/kubectl/cheatsheet/#scaling-resources
> 
> It makes much more sense to separate concerns here -
> 
> * `apply` for create/update of the ServiceSpec
> * `scale` for modifications to the PlacementSpec count
> 
> which implies `--placement` should be dropped from the `orch apply` cli ...

can we move this to a different PR?

**Comment by sebastian-philipp** [comment]
finally getting somewhere:

```
smithi139:> sudo /home/ubuntu/cephtest/cephadm --image quay.ceph.io/ceph-ci/ceph:3dc726ff39ef41323eb31a933b14a7bd46da297c shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid 2b84898e-b46f-11ea-a06d-001a4aab830c -- ceph cephadm remote daemon add osd smithi139:vg_nvme/lv_4
 -1 mgr.server reply reply (2) No such file or directory must specify host(s) to deploy on
got remote process result: 2
Saw exception from nested tasks
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/contextutil.py", line 31, in nested
    vars.append(enter())
  File "/usr/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-swagner-testing-2020-06-22-0931/qa/tasks/cephadm.py", line 646, in ceph_osds
    remote.shortname + ':' + short_dev
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-swagner-testing-2020-06-22-0931/qa/tasks/cephadm.py", line 58, in _shell_daemon_add
    'ceph', 'cephadm', 'remote', 'daemon', 'add', daemon_type, arg
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-swagner-testing-2020-06-22-0931/qa/tasks/cephadm.py", line 48, in _shell
    **kwargs
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 204, in run
    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 446, in run
    r.wait()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 160, in wait
    self._raise_for_status()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 182, in _raise_for_status
    node=self.hostname, label=self.label
teuthology.exceptions.CommandFailedError: Command failed on smithi139 with status 2: 'sudo /home/ubuntu/cephtest/cephadm --image quay.ceph.io/ceph-ci/ceph:3dc726ff39ef41323eb31a933b14a7bd46da297c shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyrin
```

**Comment by jschmid1** [comment]
let's resurrect this

**Comment by sebastian-philipp** [comment]
Hm. never managed to get this rebased.

**Comment by sebastian-philipp** [comment]
too old

**Pull request closed without merging**